### DISCUSSION TITLE
How to never show alpha/beta updates versions that are currently stable

### DISCUSSION THREAD

**Discussion by ColtonIdle** [ben-manes/gradle-versions-plugin#537] [Q&A] [answered]
I basically sort deps into two buckets.

1. Always stable, never try alpha/beta
2. Always try bleeding edge alpha/beta

This plugin out of the box is great for 2, but I do want to be able to say "Hey if I'm using a stable version, don't show me alpha/beta/etc versions" How would I do this?

**Comment by ben-manes** [comment, marked as answer]
It sounds like you'd want to use [rejectVersionsIf](https://github.com/ben-manes/gradle-versions-plugin#rejectversionsif-and-componentselection) (which is shorthand for a component selection rule). This will reject newer versions that do not match the predicate and evaluate the next version, until it finds an acceptable match. Note that an suggested `isNonStable` function is supplied in the readme, but we don't classify as such because there are many different versioning styles in the wild.

> Example 2: disallow release candidates as upgradable versions from stable versions
```groovy
tasks.named("dependencyUpdates").configure {
  rejectVersionIf {
    isNonStable(it.candidate.version) && !isNonStable(it.currentVersion)
  }
}
```

**Reply by ColtonIdle** [reply]
Thanks @ben-manes

I'm using .kts build files and added the import and snippet as per the docs

```
tasks.withType<DependencyUpdatesTask> {
    rejectVersionIf {
        isNonStable(candidate.version) && !isNonStable(currentVersion)
    }
}
```

but I get this unresolved reference error. Am I doing something wrong or are docs out of date?
<img width="588" alt="Screen Shot 2021-07-02 at 10 14 39 AM" src="https://user-images.githubusercontent.com/31751141/124287698-523a8200-db1e-11eb-8530-98d04c8bb3f2.png">

**Reply by ColtonIdle** [reply]
Didn't know that I had to copy that function. Re-read the readme. 😄 
https://github.com/ben-manes/gradle-versions-plugin#revisions

**Reply by ColtonIdle** [reply]
@ben-manes confirmed that it works! Wooo. No longer do I have a list of like 20 dependencies that I won't actually update. Thank you!

**Reply by ColtonIdle** [reply]
@ben-manes I started a new project and so I copied the above into my new project and I can't tell if it's working because it's telling me these two dependencies are out of date

```
The following dependencies have later milestone versions:
 - androidx.test.espresso:espresso-core [3.4.0 -> 3.5.0-alpha01]
     https://developer.android.com/testing
 - androidx.test.ext:junit [1.1.3 -> 1.1.4-alpha01]
     https://developer.android.com/testing
```

Any ideas why these two are getting past the filter?

**Reply by ColtonIdle** [reply]
Ah. Figured it out.

I had

```
tasks.withType<DependencyUpdatesTask> {
    rejectVersionIf { (candidate.group == "org.jacoco") && (candidate.version != currentVersion) }
    rejectVersionIf { isNonStable(candidate.version) && !isNonStable(currentVersion) }
}
```

but I need to combine them into a single rejectVersionIf statement.

**Reply by ben-manes** [reply]
nope, it should be rejected. If you use a [groovy console](http://groovyconsole.appspot.com/) and try `!('1.1.4-alpha01' ==~ /^[0-9,.v-]+(-r)?$/​)​​` then it returns true (reject). You can use `--info` to see Gradle's logs or just print during the callback.

**Reply by ben-manes** [reply]
oh congrats. Yes its not composable unfortunately.
